VAULT-40172 add docs for SCIM support (#2089)
hashicorp/web-unified-docs: content/vault/v2.x (rc)/content/docs/enterprise/scim-overview.mdx

description: >-
  Configure Vault Enterprise as a SCIM 2.0 server so identity platforms can provision and manage users and groups in Vault.
---

# Provision Vault identities with SCIM

@include 'alerts/enterprise-only.mdx'

@include 'alerts/beta.mdx'

Vault Enterprise supports System for Cross-domain Identity Management (SCIM)
2.0 as a beta feature for provisioning identity users and groups from external
identity platforms.

Vault exposes SCIM through the identity secrets engine.

- SCIM users map to Vault
  [entities](/vault/docs/concepts/identity#entities-and-aliases).
- When you configure `alias_mount_accessor`, SCIM users also create
  mount-bound [entity aliases](/vault/docs/concepts/identity#mount-bound-aliases).
- SCIM groups map to Vault
  [internal identity groups](/vault/docs/concepts/identity#external-vs-internal-groups).
- A SCIM client can only see and manage the users and groups it created.
- SCIM does not manage Vault policies.

Use this guide to enable SCIM, create a SCIM client, authenticate that client,
and send SCIM requests to Vault.

## Before you start

- Use Vault Enterprise.
- Permissions to enable auth methods and manage identity resources.

> - Choose an auth method for the SCIM client.

Reviewer: ```suggestion
- You should know the auth method you want to use for the SCIM client.
```

If you use Vault Enterprise namespaces, you can address SCIM in either of these
ways:

- Put the namespace in the request path, such as
  `/v1/admin/identity/scim/v2/Users`.
- Use the standard Vault namespace header with the non-namespaced path.

## Configure a SCIM client

Create a SCIM client before you send SCIM protocol requests.

Each SCIM client represents one external provisioning system. The SCIM client
configuration consists of these fields:

- `client_name` (part of the URL): Namespace-scoped name for the external system.
- `access_grant_principal`: ID of the Vault entity that authorizes SCIM API
  requests for that client.
- `alias_mount_accessor`: Optional auth mount accessor. When set, Vault also
  creates a mount-bound entity alias for each SCIM user.

Vault enforces these rules:

- `access_grant_principal` must refer to an existing entity in the same
  namespace.
- `alias_mount_accessor`, when set, must refer to an existing non-local auth
  mount in the same namespace.
- `alias_mount_accessor` is immutable after creation.
- A SCIM client can only manage the users and groups it created.

~> **Note:** Currently `alias_mount_accessor` only supports non-local

> auth mounts. This limitation will be removed in the future.

Reviewer: ```suggestion
Vault currently limit `alias_mount_accessor` supports to non-local
auth mounts.
```
Style correction: avoid folding normal content into aside, focus on current functionality (don't make promises about future functionality)

```shell-session
$ vault write identity/scim/client/example-client \
    access_grant_principal="<entity_id>"
```

If you want Vault to create aliases for provisioned users, add
`alias_mount_accessor="<auth_mount_accessor>"` to this command.

### 4. Create an entity alias on the token auth mount

List auth mounts and record the accessor for `token/`.

```shell-session
$ vault auth list
```

Create an alias on that mount for the SCIM client entity.

```shell-session
$ vault write identity/entity-alias \
    name="example-client" \
    mount_accessor="<token_mount_accessor>" \
    canonical_id="<entity_id>"
```

### 5. Create a token role and mint a bearer token

```shell-session
$ vault write auth/token/roles/example-scim \
    allowed_entity_aliases="example-client" \
    orphan=true \
    token_no_default_policy=true

$ vault token create \
    -role=example-scim \
    -entity-alias=example-client \
    -policy=scim-client \
    -no-default-policy
```

Use the returned token value as the bearer token for SCIM requests.

~> **Note:** Plan token renewal or rotation around the TTL configured on
your token auth mount.

## Send SCIM requests

> Use the Vault token in the request header.

Reviewer: ```suggestion
Use the Vault token in the request header to call the SCIM endpoint:
```
